server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/identity/PostgresCustomIdentityDAO.java

```diff
@@ -0,0 +1,233 @@
+                .set(BCC, convertToStringArray(identity.bcc()))
+                .onConflict(USERNAME, ID)
+                .doUpdate()
+                .set(NAME, identity.name())
+                .set(EMAIL, identity.email().asString())
+                .set(TEXT_SIGNATURE, identity.textSignature())
+                .set(HTML_SIGNATURE, identity.htmlSignature())
+                .set(MAY_DELETE, identity.mayDelete())
+                .set(SORT_ORDER, identity.sortOrder())
+                .set(REPLY_TO, convertToStringArray(identity.replyTo()))
+                .set(BCC, convertToStringArray(identity.bcc()))))
+            .thenReturn(identity);
+    }
+
+    @Override
+    public Publisher<BoxedUnit> delete(Username username, Seq<IdentityId> ids) {
+        return executorFactory.create(username.getDomainPart())
+            .executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+                .where(USERNAME.eq(username.asString()))
+                .and(ID.in(JavaConverters.asJavaCollection(ids).stream().map(IdentityId::id).collect(ImmutableList.toImmutableList())))))
+            .thenReturn(BoxedUnit.UNIT);
+    }
+
+    @Override
+    public Publisher<BoxedUnit> delete(Username username) {
+        return executorFactory.create(username.getDomainPart())
+            .executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+                .where(USERNAME.eq(username.asString()))))
+            .thenReturn(BoxedUnit.UNIT);
+    }
+
+    private Identity readRecord(Record record) throws Exception {
+        return new Identity(new IdentityId(record.get(ID)),
+            record.get(SORT_ORDER),
+            record.get(NAME),
+            new MailAddress(record.get(EMAIL)),
+            convertToScala(record.get(REPLY_TO)),
+            convertToScala(record.get(BCC)),
+            record.get(TEXT_SIGNATURE),
+            record.get(HTML_SIGNATURE),
+            record.get(MAY_DELETE));
+    }
+
+    private Option<List<EmailAddress>> convertToScala(String[] jsonArray) {
+        return OptionConverters.toScala(Optional.of(JavaConverters.asScalaIterator(Arrays.stream(jsonArray)
+                .map(Throwing.function(this::convertToObject))
+                .map(Throwing.function(email -> EmailAddress.from(Optional.ofNullable(email.getName()), new MailAddress(email.getEmail()))))
+                .iterator())
+            .toList()));
+    }
+
+    private String[] convertToStringArray(Option<List<EmailAddress>> maybeEmailAddresses) {
+        return OptionConverters.toJava(maybeEmailAddresses).map(emailAddresses ->
+                JavaConverters.asJavaCollection(emailAddresses).stream()
+                    .map(emailAddress -> convertToJson(new Email(emailAddress.nameAsString(),
+                        emailAddress.email().asString()))).toArray(String[]::new))
+            .orElse(new String[0]);
+    }
+
+    private String convertToJson(Email email) {
```

Comment:
Why we don't use Postgresql `JSON` type directly?
Instead of `convertToJson` (asString) -> and save to columm `Array String`

Reply:
In cassandra, it is the array of udt. Therefore, I think string array in Postgres is equivalent. array of json is not supported by JOOQ. array json does not fit very well I think

Reply:
Json of list?

Reply:
uh

Reply:
My meaning is we can serialize/ deserialize directly `Json of list`, we don't need array json.

Reply:
What is the difference when using Json of list? I think the only difference is that Postgres check if it is valid json

Reply:
> array json (the whole array in json format) does not fit very well I think

Just curious but why do you feel like that? 
IMO storing an array in a JSON would be better.  If we work and store data under JSON, we should use JSONB to enforce JSON validation. (we did the same in other PRs).
Storing JSON under Array[String] can not prevent inserting malformed strings and leads to deserialization issues.

Reply:
Regarding validation, json library (jackson, ...) has already validated strings

Reply:
Another story like mutable/ immutable objects for example.

We use a mutable object and we are confident that we would not change that object in the middle. But who knows?

That is why we follow best practices to avoid using mutable objects for example.

Reply:
well ok If you guys insist on that, I will change

Reply:
updated